src/Analyzers/CSharp/Analyzers/FileBasedPrograms/AppDirectiveHelpers.cs

```diff
@@ -0,0 +1,705 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// Largely copied from https://github.com/dotnet/sdk/blob/ce691a53dad7f2b6057c85c1e7fd7800dfe0788f/src/Cli/dotnet/Commands/Run/VirtualProjectBuildingCommand.cs#L1410
+// TODO: we'd like to extract a source package which makes this copy of the source unnecessary.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Xml;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.FileBasedPrograms;
+
+internal static class AppDirectiveHelpers
+{
+
+#pragma warning disable RSEXPERIMENTAL003 // 'SyntaxTokenParser' is experimental
```

Comment:
I think SyntaxTokenParser is not experimental anymore (but in the sdk we don't have that new version of roslyn yet)